app/src/main/java/com/example/tasky/feature_login/presentation/LoginViewModel.kt

```diff
@@ -16,27 +55,100 @@ class LoginViewModel : ViewModel() {
+    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000L), false)
 
     fun onNameChange(newName: String) {
         _name.value = newName
-        _isNameValid.value = newName.isValidName()
     }
 
     fun onEmailChange(newEmail: String) {
         _email.value = newEmail
-        _isEmailValid.value = newEmail.isValidEmail()
     }
 
     fun onPasswordChange(newPassword: String) {
         _password.value = newPassword
-        _isPasswordValid.value = newPassword.isValidPassword()
+    }
+
+    fun onPasswordVisibilityClick() {
+        _isPasswordVisible.value = !isPasswordVisible.value
+    }
+
+    fun logInClicked() {
+        Log.d(
+            TAG,
+            "logInClicked and userCredentials is email: ${email.value} and password: ${password.value}"
+        )
+        viewModelScope.launch {
+            _viewState.emit(AuthenticationViewState.Loading)
+            val result = userRemoteRepository.logInUser(LoginUserInfo(email.value, password.value))
+
+            when (result) {
+                is Result.Success -> {
+                    println("success login!")
+                    // emit a viewState to change to agenda composable
+                    _viewState.emit(AuthenticationViewState.Success)
+                }
+
+                is Result.Error -> {
+                    println("failed login :(")
+                    // emit a viewState to show ErrorMessage
+                    val errorMsg = result.error.formatErrorMessage()
+                    _viewState.emit(AuthenticationViewState.Failure(errorMsg))
+                }
+            }
+        }
+    }
+
+    fun registerUserClicked() {
+        Log.d(
+            TAG,
+            "registerUserClicked and userInfo is name: ${name.value}, email: ${email.value}, password: ${password.value}"
+        )
+        viewModelScope.launch {
+            _viewState.emit(AuthenticationViewState.Loading)
+            val result = userRemoteRepository.registerUser(
+                RegisterUserInfo(
+                    name.value,
+                    email.value,
+                    password.value
+                )
+            )
+
+            when (result) {
+                is Result.Success -> {
+                    println("success register!")
+                    // emit a viewState to show LoginScreen composable
+                    _viewState.emit(AuthenticationViewState.Success)
+                }
+
+                is Result.Error -> {
+                    println("failed register :(")
+                    // emit a viewState to show ErrorMessage
+                    val errorMsg = result.error.formatErrorMessage()
+                    _viewState.emit(AuthenticationViewState.Failure(errorMsg))
+                }
+            }
+        }
+    }
+
+    fun onShowErrorDialog(message: String) {
+        _uiState.value = _uiState.value.copy(showErrorDialog = true, dialogMessage = message)
```

Comment:
In general, I'd always use `_stateFlow.update { it.copy(...) }` to avoid race conditions. This is how it could happen:

*StateFlow holds value A*
- Thread 1 calls `copy()` on StateFlow with value A
- Before Thread 1 assigns its changed copy, Thread 2 updates StateFlow with value B
- Now, Thread 1 assigns its changed copy, which is still based on value A though, but the latest value is already value B. Therefore, value B would never be considered here.

Not always an issue, but I worked on highly stateful apps with frequently changing states, in which this really bit me, since such bugs are super hard to find.
Using `_state.value = _state.value.copy(...)` is fine if you always call this from a single thread (which you do here). But in ViewModels, you're very likely to not do that as soon as you'd change the state inside a `viewModelScope`.